### ISSUE TITLE
Priority example not executing steps in desired order

### ISSUE THREAD

**Issue by berkowski** [rtfm-rs/cortex-m-rtfm#108]
I'm working my way through the RTFM book, doing the examples as I go.   When I got to the Priorities section the example didn't seem to run as outlined.  This was my own code, but if I try the example (`lock.rs`) in the repo:

```
$cargo run --example lock
   Compiling semver-parser v0.7.0
   Compiling proc-macro2 v0.4.24                                                             
   Compiling unicode-xid v0.1.0                                                                        
   Compiling rand_core v0.3.0                                                                          
   Compiling typenum v1.10.0                                                                           
   Compiling cortex-m v0.5.8                                                                           
   Compiling vcell v0.1.0                                                                              
   Compiling cortex-m-semihosting v0.3.2                                                               
   Compiling cortex-m-rt v0.6.5                                                                        
   Compiling aligned v0.2.0                                                                            
   Compiling byteorder v1.2.7                                                                          
   Compiling lm3s6965 v0.1.3                                                                           
   Compiling r0 v0.2.2                                                                                 
   Compiling cortex-m-rtfm v0.4.0-beta.3 (/home/rust/cortex-m-rtfm)                
   Compiling stable_deref_trait v1.1.1                                                                 
   Compiling panic-halt v0.2.0                                                                         
   Compiling volatile-register v0.2.0                                                                  
   Compiling rand_core v0.2.2                                                                          
   Compiling semver v0.9.0                                                                             
   Compiling rand v0.5.5                                                                               
   Compiling hash32 v0.1.0                                                                             
   Compiling rustc_version v0.2.3                                                                      
   Compiling bare-metal v0.2.4                                                                         
   Compiling generic-array v0.11.1                                                                     
   Compiling generic-array v0.12.0                                                                     
   Compiling quote v0.6.10                                                                             
   Compiling heapless v0.4.0                                                                           
   Compiling as-slice v0.1.0                                                                           
   Compiling panic-semihosting v0.5.1                                                                  
   Compiling syn v0.15.22                                                                              
   Compiling owned-singleton-macros v0.1.0                                                             
   Compiling cortex-m-rt-macros v0.1.3                                                                
   Compiling cortex-m-rtfm-macros v0.4.0-beta.3 (/home/rust/cortex-m-rtfm/macros)  
   Compiling owned-singleton v0.1.0                                                                    
   Compiling alloc-singleton v0.1.0                                                                    
    Finished dev [unoptimized + debuginfo] target(s) in 18.83s                                         
     Running `qemu-system-arm -cpu cortex-m3 -machine lm3s6965evb -nographic -semihosting-config enable=on,target=native -kernel target/thumbv7m-none-eabi/debug/examples/lock`
A
D - SHARED = 2
B - SHARED = 2
C
E
```
Similar results when compiled with `--release`.  I'm using `rustc 1.31.0-beta.22 (d2e3b5332 2018-12-01)`

**Comment by japaric** [comment]
This is ... interesting. We run the `lock` example as part of the CI and verify that its output matches [our expectations](https://github.com/japaric/cortex-m-rtfm/blob/c2fbb2848851e32b78e79ff9e919538b7d5ab8a0/ci/expected/lock.run) for both nightly and stable.

Could you try running the example from the `master` branch using a stable compiler to see if you still get the wrong output?

**Comment by berkowski** [comment]
Different machine (same OS, Ubuntu 16.04 LTS):
```
$ git describe
v0.4.0-beta.2-26-g3511e91

$ qemu-system-arm --version
QEMU emulator version 2.5.0 (Debian 1:2.5+dfsg-5ubuntu10.33), Copyright (c) 2003-2008 Fabrice Bellard

$ uname -a
Linux spiff 4.15.0-42-generic #45~16.04.1-Ubuntu SMP Mon Nov 19 13:02:27 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

$ cargo clean
$ cargo run --example lock
   Compiling semver-parser v0.7.0                                                                        
   Compiling proc-macro2 v0.4.24                                                                         
   Compiling unicode-xid v0.1.0                                                                          
   Compiling rand_core v0.3.0                                                                            
   Compiling typenum v1.10.0                                                                             
   Compiling vcell v0.1.0                                                                                
   Compiling cortex-m v0.5.8                                                                             
   Compiling byteorder v1.2.7                                                                            
   Compiling aligned v0.2.0                                                                              
   Compiling cortex-m-semihosting v0.3.2                                                                 
   Compiling cortex-m-rt v0.6.7                                                                          
   Compiling cortex-m-rtfm v0.4.0 (/home/zac/repos/rust/cortex-m-rtfm)                                   
   Compiling stable_deref_trait v1.1.1                                                                   
   Compiling r0 v0.2.2                                                                                   
   Compiling lm3s6965 v0.1.3                                                                             
   Compiling panic-halt v0.2.0                                                                           
   Compiling volatile-register v0.2.0                                                                    
   Compiling rand_core v0.2.2                                                                            
   Compiling hash32 v0.1.0                                                                               
   Compiling semver v0.9.0                                                                               
   Compiling rand v0.5.5                                                                                 
   Compiling rustc_version v0.2.3                                                                        
   Compiling bare-metal v0.2.4                                                                           
   Compiling generic-array v0.11.1                                                                       
   Compiling generic-array v0.12.0                                                                       
   Compiling quote v0.6.10                                                                               
   Compiling as-slice v0.1.0                                                                             
   Compiling heapless v0.4.1                                                                             
   Compiling syn v0.15.23                                                                                
   Compiling panic-semihosting v0.5.1                                                                    
   Compiling cortex-m-rt-macros v0.1.5                                                                   
   Compiling owned-singleton-macros v0.1.0                                                               
   Compiling cortex-m-rtfm-macros v0.4.0 (/home/zac/repos/rust/cortex-m-rtfm/macros)                     
   Compiling owned-singleton v0.1.0                                                                      
   Compiling alloc-singleton v0.1.0                                                                      
    Finished dev [unoptimized + debuginfo] target(s) in 13.92s                                           
     Running `qemu-system-arm -cpu cortex-m3 -machine lm3s6965evb -nographic -semihosting-config enable=on,target=native -kernel target/thumbv7m-none-eabi/debug/examples/lock`
A
D - SHARED = 2
B - SHARED = 2
C
E
```

Happy to help track down what's going on, even if it's just a misconception on my part.  I've really enjoyed watching the embedded scene develop with rust but I'm not the best with the tools.

**Comment by japaric** [comment]
I looked into this and I can repro the problem when I use `qemu-system-arm` v2.5.0 but not when I use v3.1.0 -- I used the same Rust binary in both cases. I assume that BASEPRI is not properly implemented in older version of QEMU; I couldn't find any changelog that said that BASEPRI was recently implemented though.

I would suggest trying a new QEMU. If you want to dig into this further problem with QEMU 2.5.0 I would suggest printing the value of BASEPRI (`cortex_m::register::basepri::read`) outside and inside of the `lock` closures.
